### ISSUE TITLE
Add atomic increment/decrement operations to IDistributedCache

### ISSUE THREAD

**Issue by khellang** [dotnet/runtime#36386]
It would be really nice to have for scenarios like rate limiting.

I don't know about SQL Server support for this, but I guess this would map to Redis' `INCR` and `DECR` operations.
If the key does not exist, it would be set to 0 before performing the operation. The methods would return the value of key after increment/decrement.

I propose the following new members:

```diff
namespace Microsoft.Extensions.Caching.Distributed
{
    public interface IDistributedCache
    {
        // Existing members:
        byte[] Get(string key);
        Task<byte[]> GetAsync(string key, CancellationToken token = default(CancellationToken));
        void Set(string key, byte[] value, DistributedCacheEntryOptions options);
        Task SetAsync(string key, byte[] value, DistributedCacheEntryOptions options, CancellationToken token = default(CancellationToken));
        void Refresh(string key);
        Task RefreshAsync(string key, CancellationToken token = default(CancellationToken));
        void Remove(string key);
        Task RemoveAsync(string key, CancellationToken token = default(CancellationToken));

        // New members:
+       long Increment(string key, DistributedCacheEntryOptions options);
+       Task<long> IncrementAsync(string key, DistributedCacheEntryOptions options, CancellationToken token = default(CancellationToken));
+       long Decrement(string key, DistributedCacheEntryOptions options);
+       Task<long> DecrementAsync(string key, DistributedCacheEntryOptions options, CancellationToken token = default(CancellationToken));
+       long IncrementBy(string key, long value, DistributedCacheEntryOptions options);
+       Task<long> IncrementByAsync(string key, long value, DistributedCacheEntryOptions options, CancellationToken token = default(CancellationToken));
+       long DecrementBy(string key, long value, DistributedCacheEntryOptions options);
+       Task<long> DecrementByAsync(string key, long value, DistributedCacheEntryOptions options, CancellationToken token = default(CancellationToken));
    }
}
```

@Eilon @Tratcher @davidfowl Is this something you'd consider for 3.0?

**Comment by glucaci** [comment]
there are any news for this feature ?

**Comment by khellang** [comment]
Yeah, I'd still like to get this feature in. What would be the best way to attack this? Submit a PR for an additional interface with the new members? Does it have to go through API review?

**Comment by davidfowl** [comment]
Everything has to go through API review and it's a breaking change and we haven't even agreed that this makes sense on anything besides a redis implementation.

**Comment by khellang** [comment]
> it's a breaking change

Adding a separate interface that implementations can opt into is a breaking change?

> we haven't even agreed that this makes sense on anything besides a redis implementation.

Most other platforms seems to be able to handle this just fine, whether the backing store is in-memory, Redis, Memcached or SQL-based 🤷‍♂️ 

What can we do to further this proposal?

**Comment by eerhardt** [comment]
> What can we do to further this proposal?

Check out the API proposal review process here: https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md. Specifically this part:

> Please use [this template](https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.md&title=). The issue should have the label api-suggestion. Here is a [good example](https://github.com/dotnet/runtime/issues/38344) of an issue following that template.

The current proposal on top of this issue would be a breaking change, which we've decided we can't take. To make progress, can you edit the top proposal to match the template with the newly proposed APIs?

**Comment by YohanSciubukgian** [comment]
Could it be build on top of efcore?
You won’t have to manage/choose the backing store to use.
It will also let developpers choose the right backing store (inmemory, redis, sql, cosmos, etc...) based on their usages.

**Comment by atuls9390** [comment]
@khellang Any work-around for this?

**Comment by khellang** [comment]
Not really. I ended up just using the Redis API directly and not shipping a library to NuGet as it wouldn't be as useful without proper abstractions.

**Comment by pinkfloydx33** [comment]
I ended up having to switch my abstractions over to the replacements offered by [Foundatio](https://github.com/FoundatioFx/Foundatio.Redis#caching) just to get increment/decrement. I'd prefer to use those offered by the framework rather than leek Foundatio into all my projects (though I admit there's definite bonus points for all the other stuff their abstractions offer.).

**Comment by AlOnestone01** [comment]
Hi. Are there any updates on this? We would also love to see this being implemented!

**Comment by rosebyte** [comment]
Triage: although we like these methods, we think they belong to a different interface and likely different library.
